### DISCUSSION TITLE
Dynamic Code Evaluation (e. g. 'eval', 'new Function') not allowed in Middleware pages/_middleware

### DISCUSSION THREAD

**Discussion by rajesh40** [vercel/next.js#37038] [Help]
Dynamic Code Evaluation (e. g. 'eval', 'new Function') not allowed in Middleware pages/_middleware

We made a middleware, it worked as expected but building giving error like 
Dynamic Code Evaluation (e. g. 'eval', 'new Function') not allowed in Middleware pages/_middleware

**Comment by icyJoseph** [comment]
Hi,

The middleware uses a different API, some might say limited, in order to be portable, and fast. These are the supported APIs, https://nextjs.org/docs/api-reference/edge-runtime#runtime-apis. 

Every iteration the team is getting better and better at catching libraries that use things that are not well supported, or won't be supported at all, during dev mode, or build time.

Some libraries make use of certain techniques that wouldn't work well on an edge environment. 

In this case, it'd help a lot if you could share more of the error, as it surely says which library is at fault here. Recently a project of mine had to stop using Pino to log in the middleware, and instead roll our own solution. 

I am surprised you managed to even work on dev mode with the middleware at fault, in 12.1.6, we couldn't even get the application started, without problems coming from a deep dependency of Pino logger.

Again, if you could share more of the logs/error, so perhaps we can give an alternative library. For instance for crypto stuff, we are using `jose`, which uses the basic crypto API available to expose various algorithms.

> Also be mindful that middleware is at beta stage, and things are still developing. In parallel, the JavaScript ecosystem is inching closer and closer to universal JavaScript, by writing libraries that make no assumptions about the environment they run in.
